src/lib/apiTypesV2/Entity.cpp

```diff
@@ -61,6 +62,128 @@ Entity::~Entity()
+*   attributeVector preparation. Thus, either we leave the function heare and mongoBackend calls
+*   Entity::filterAttributes() or the logic is moved to mongoBackend (to be decided when we face
+*   this FIXME).
+*/
+void Entity::filterAttributes
+(
+  const std::vector<std::string>&  attrsFilter,
+  bool                             dateCreatedOption,
+  bool                             dateModifiedOption
+)
+{
+  bool dateCreatedAdded  = false;
+  bool dateModifiedAdded = false;
+
+  if (attrsFilter.size () != 0)
+  {
+    if (std::find(attrsFilter.begin(), attrsFilter.end(), ALL_ATTRS) != attrsFilter.end())
+    {
+      // Not filtering, just adding dateCreated and dateModified if needed
+      if ((creDate != 0) && (std::find(attrsFilter.begin(), attrsFilter.end(), DATE_CREATED) != attrsFilter.end()))
+      {
+        ContextAttribute* caP = new ContextAttribute(DATE_CREATED, DATE_TYPE, creDate);
+        attributeVector.push_back(caP);
+        dateCreatedAdded = true;
+      }
+      if ((modDate != 0) &&  (std::find(attrsFilter.begin(), attrsFilter.end(), DATE_MODIFIED) != attrsFilter.end()))
+      {
+        ContextAttribute* caP = new ContextAttribute(DATE_MODIFIED, DATE_TYPE, modDate);
+        attributeVector.push_back(caP);
+        dateModifiedAdded = true;
+      }
+    }
+    else
+    {
+      // Reorder attributes in the same order they are in attrsFilter, excluding the ones
+      // note there (i.e. filtering them out) and giving special treatment to creation
+      // and modification dates
+      //
+      // The (attributeVector.lookup(DATE_XXXX) == -1) check is to give preference to user
+      // defined attributes (see
+      // https://fiware-orion.readthedocs.io/en/master/user/ngsiv2_implementation_notes/index.html#datemodified-and-datecreated-attributes)
+
+      std::vector<ContextAttribute*> caNewV;
+
+      for (unsigned int ix = 0; ix < attrsFilter.size(); ix++)
+      {
+        std::string attrsFilterItem = attrsFilter[ix];
+        if ((creDate != 0) && (attrsFilterItem == DATE_CREATED) && (attributeVector.lookup(DATE_CREATED) == -1))
+        {
+          ContextAttribute* caP = new ContextAttribute(DATE_CREATED, DATE_TYPE, creDate);
+          caNewV.push_back(caP);
+          dateCreatedAdded = true;
+        }
+        else if ((modDate != 0) && (attrsFilterItem == DATE_MODIFIED) && (attributeVector.lookup(DATE_MODIFIED) == -1))
+        {
+          ContextAttribute* caP = new ContextAttribute(DATE_MODIFIED, DATE_TYPE, modDate);
+          caNewV.push_back(caP);
+          dateModifiedAdded = true;
+        }
+        // Actual attribute filtering only takes place if '*' was not used
+        else
+        {
+          int found = attributeVector.lookup(attrsFilterItem);
+          if (found != -1)
+          {
+            caNewV.push_back(attributeVector.vec[found]);
+            attributeVector.vec.erase(attributeVector.vec.begin() + found);
```

Comment:
Is this necessary?
Later, "attributeVector.release();" should take care of all of it, right?
￼
￼

Reply:
If we don't erase() from the original vector, then the same caP will be pointed from two places (the old vector and the new vector) so it will freed twice, causing a double free.

I have a valgrind run in place to confirm if there is something wrong with regards to memory management in the current implementation.

Reply:
ok, I guess ...
NYC